src/shell/filesystem_shell.rs

```diff
@@ -95,132 +96,46 @@ impl Shell for FilesystemShell {
-            let entry = Path::new(&full_path);
-            if entry.is_dir() {
-                let entries = std::fs::read_dir(&entry);
-                let entries = match entries {
-                    Err(e) => {
-                        if let Some(s) = path {
-                            return Err(ShellError::labeled_error(
-                                e.to_string(),
-                                e.to_string(),
-                                s.tag(),
-                            ));
-                        } else {
-                            return Err(ShellError::labeled_error(
-                                e.to_string(),
-                                e.to_string(),
-                                name_tag,
-                            ));
-                        }
-                    }
-                    Ok(o) => o,
-                };
-                let mut entries = entries.collect::<Vec<Result<std::fs::DirEntry, _>>>();
-                entries.sort_by(|x, y| match (x, y) {
-                    (Ok(entry1), Ok(entry2)) => entry1
-                        .path()
-                        .to_string_lossy()
-                        .to_lowercase()
-                        .cmp(&entry2.path().to_string_lossy().to_lowercase()),
-                    (Err(_), Ok(_)) => std::cmp::Ordering::Greater,
-                    (Ok(_), Err(_)) => std::cmp::Ordering::Greater,
-                    _ => std::cmp::Ordering::Equal,
-                });
-                let stream = async_stream! {
-                    for entry in entries {
-                        if ctrl_c.load(Ordering::SeqCst) {
-                            break;
-                        }
-                        if let Ok(entry) = entry {
-                            let filepath = entry.path();
-                            if let Ok(metadata) = std::fs::symlink_metadata(&filepath) {
-                                let mut filename = if let Ok(fname) = filepath.strip_prefix(&cwd) {
-                                    fname
-                                } else {
-                                    Path::new(&filepath)
-                                };
-
-                                if short_names {
-                                    filename = if let Some(fname) = filename.file_name() {
-                                        Path::new(fname)
-                                    } else {
-                                        let fname = filename
-                                            .file_name()
-                                            .or_else(|| Path::new(filename).file_name())
-                                            .unwrap_or(filename.as_os_str());
-                                        Path::new(fname)
-                                    }
-                                }
+        let paths: Vec<GlobResult> = match glob::glob(&path.to_string_lossy()) {
```

Comment:
If we're switching to the `glob` crate, I'd recommend sticking with an iterator. In other words, no `collect`. That way we can stream the results (always try to stream! 🙂)

Reply:
Is there a way to otherwise check the first result to see if the iterator is empty?

Reply:
Great question. You can make an iterator "peekable" (see https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.peekable), which gives it a `peek` method where you can check the next item without actually advancing the iterator.

Reply:
Is there a downside to a peekable iterator seems like that should be the default or something.

Reply:
I think the downside is that there's a small amount of cost involved in maintaining the iterator in such a way. Rust is all about "don't pay for what you don't use". If I had to guess, that's probably why!